### PR TITLE
refactor(filter): make it generic

### PR DESCRIPTION
**Pull request by sileht**
This makes the Filter class Generic.

The binary implementation is now called FilterBinary

A future implementation will be done to return the more recent
timestamp.

Change-Id: Ice711d1f106d445badf6dc5539fc1b309887b35a

### CODE REVIEW

**Comment by sileht** [comment]
This pull request is part of a stack:
* feat(rules): allow and/or in conditions ([#2691](https://github.com/Mergifyio/mergify-engine/pull/2691))
* feat: allow merged/closed/author/base/head in and/or conditions ([#2692](https://github.com/Mergifyio/mergify-engine/pull/2692))
* feat(conditions): allow checks condition in and/or ([#2693](https://github.com/Mergifyio/mergify-engine/pull/2693))
* refactor(filter): make it generic ([#2694](https://github.com/Mergifyio/mergify-engine/pull/2694))
* feat(time) add time conditions ([#2695](https://github.com/Mergifyio/mergify-engine/pull/2695))

File: .github/workflows/ci.yaml [hunk 1]
```diff
@@ -89,6 +89,14 @@ jobs:
...
           pip install tox
           tox -e docs
+      # NOTE(sileht): workflow run on/pull_request doesn't have access to
+      # secrets
+      # So, we upload it as artefact, then another workflow docs-pr.yaml will
+      # upload it on gcp.
+      - uses: actions/upload-artifact@v2
+        with:
+          name: docs-pr-${{ github.event.pull_request.number }}
+          path: docs/build
```
Comment (jd) [inline .github/workflows/ci.yaml:99]: ?

File: .github/workflows/docs-pr.yaml [hunk 2]
```diff
@@ -0,0 +1,48 @@
+name: Upload temporary docs of pull request to GCP
```
Comment (jd) [inline .github/workflows/docs-pr.yaml:1]: ditto

**Review by jd — CHANGES_REQUESTED** [review]

File: mergify_engine/actions/merge_base.py [hunk 3]
```diff
@@ -99,6 +99,64 @@ def strict_merge_parameter(v):
...
+        "something unrelated to checks doesn't match? %s",
+        conditions_without_checks.get_summary(),
+    )
+    if not conditions_without_checks.match:
+        return check_api.Conclusion.FAILURE
+
+    # NOTE(sileht): Have all checks reported their status?
+    await conditions_with_all_checks(ctxt.pull_request)
+    ctxt.log.debug(
+        "are checks reported their status? %s",
```
Comment (jd) [inline mergify_engine/actions/merge_base.py:142]:
```suggestion
        "did check report their status? %s",
```

File: mergify_engine/github_events.py [hunk 4]
```diff
@@ -53,7 +53,13 @@ def meter_event(
...
 
 def _extract_slim_event(event_type, data):
-    slim_data = {"sender": data["sender"]}
+    slim_data = {
+        "sender": {
+            "id": data["sender"]["id"],
+            "login": data["sender"]["login"],
+            "type": data["sender"]["type"],
+        }
+    }
```
Comment (jd) [inline mergify_engine/github_events.py:62]: I feel like we merged this somewhere else

**Review by jd — DISMISSED** [review]

File: .github/workflows/ci.yaml (hunk 1 above)
Comment (sileht) [reply .github/workflows/ci.yaml]: I don't get why this show up here, I just rebase/push and it disapears.

**Review by jd — APPROVED** [review]

**Pull request merged**